libcxx/include/__expected/expected.h

```diff
@@ -88,8 +88,289 @@ _LIBCPP_HIDE_FROM_ABI void __throw_bad_expected_access(_Arg&& __arg) {
+               is_trivially_copy_constructible_v<_Tp> && is_trivially_copy_constructible_v<_Err>)
+    = default;
+    _LIBCPP_HIDE_FROM_ABI constexpr __repr(__repr&&) = delete;
+    _LIBCPP_HIDE_FROM_ABI constexpr __repr(__repr&&)
+      requires(is_move_constructible_v<_Tp> && is_move_constructible_v<_Err> &&
+               is_trivially_move_constructible_v<_Tp> && is_trivially_move_constructible_v<_Err>)
+    = default;
+
+    _LIBCPP_HIDE_FROM_ABI constexpr __repr& operator=(const __repr&) = delete;
+    _LIBCPP_HIDE_FROM_ABI constexpr __repr& operator=(__repr&&)      = delete;
+
+    _LIBCPP_HIDE_FROM_ABI constexpr ~__repr()
+      requires(is_trivially_destructible_v<_Tp> && is_trivially_destructible_v<_Err>)
+    = default;
+
+    _LIBCPP_HIDE_FROM_ABI constexpr ~__repr()
+      requires(!is_trivially_destructible_v<_Tp> || !is_trivially_destructible_v<_Err>)
+    {
+      __destroy_union_member();
+    }
+
+    _LIBCPP_HIDE_FROM_ABI constexpr void __destroy_union()
+      requires(!__put_flag_in_tail && (is_trivially_destructible_v<_Tp> && is_trivially_destructible_v<_Err>))
+    {
+      std::destroy_at(&__union_.__v);
+    }
+
+    _LIBCPP_HIDE_FROM_ABI constexpr void __destroy_union()
+      requires(!__put_flag_in_tail && (!is_trivially_destructible_v<_Tp> || !is_trivially_destructible_v<_Err>))
+    {
+      __destroy_union_member();
+      std::destroy_at(&__union_.__v);
+    }
+
+    template <class... _Args>
+    _LIBCPP_HIDE_FROM_ABI constexpr void __construct_union(in_place_t, _Args&&... __args)
+      requires(!__put_flag_in_tail)
+    {
+      std::construct_at(&__union_.__v, in_place, std::forward<_Args>(__args)...);
+      __has_val_ = true;
+    }
+
+    template <class... _Args>
+    _LIBCPP_HIDE_FROM_ABI constexpr void __construct_union(unexpect_t, _Args&&... __args)
+      requires(!__put_flag_in_tail)
+    {
+      std::construct_at(&__union_.__v, unexpect, std::forward<_Args>(__args)...);
+      __has_val_ = false;
+    }
+
+  private:
+    template <class, class>
+    friend class __expected_base;
+
+    _LIBCPP_HIDE_FROM_ABI constexpr void __destroy_union_member()
+      requires(!is_trivially_destructible_v<_Tp> || !is_trivially_destructible_v<_Err>)
+    {
+      if (__has_val_) {
+        std::destroy_at(std::addressof(__union_.__v.__val_));
+      } else {
+        std::destroy_at(std::addressof(__union_.__v.__unex_));
+      }
+    }
+
+    template <class _OtherUnion>
+    _LIBCPP_HIDE_FROM_ABI static constexpr __union_t __make_union(bool __has_val, _OtherUnion&& __other)
+      requires(!__put_flag_in_tail)
+    {
+      if (__has_val)
+        return __union_t(in_place, std::forward<_OtherUnion>(__other).__val_);
+      else
+        return __union_t(unexpect, std::forward<_OtherUnion>(__other).__unex_);
+    }
+
+    _LIBCPP_NO_UNIQUE_ADDRESS __conditional_no_unique_address<__put_flag_in_tail, __union_t> __union_;
+    _LIBCPP_NO_UNIQUE_ADDRESS bool __has_val_;
```

Comment:
I think this application of `__conditional_no_unique_address` is not necessary. Indeed, if there is tail padding available after the `__union_t`, you mark it as `[[no_unique_address]]`, which is okay. If there is *no* padding, you don't mark it as `[[no_unique_address]]`, but in reality you could mark it because it just wouldn't change anything (since there are no padding bytes to reuse). After discussing this with @huixie90, we think you could simplify the overall layout of the class like so:

```
template <class _Tp, class _Err>
class expected {
  union __union_t {
    // ...ctors...

    _LIBCPP_NO_UNIQUE_ADDRESS _Tp __val_;
    _LIBCPP_NO_UNIQUE_ADDRESS _Err __unex_;
  };

  struct __repr {
    // ...ctors...

    _LIBCPP_NO_UNIQUE_ADDRESS __union_t __union_; // NOTE: always no_unique_address
    _LIBCPP_NO_UNIQUE_ADDRESS bool __has_val_;
  };

  static constexpr bool __allow_reusing_expected_tail_padding = sizeof(__union_t) - __libcpp_datasizeof<__union_t>::value <= sizeof(bool);
  _LIBCPP_NO_UNIQUE_ADDRESS __conditional_no_unique_address<__allow_reusing_expected_tail_padding, __repr> __repr_;
};
```

By using `__libcpp_datasizeof`, we can get rid of `__fits_in_tail_padding` in this patch and IMO it makes the logic a bit easier to understand. What do you think?

Essentially, the idea here is that we make everything `[[no_unique_address]]` all the time, except the top-level repr in case there are padding bytes left in the `__union_t` part of the `expected`, since the `__union_t` is the one we use `construct_at` on. This is very similar (in fact it's equivalent) to what you have right now, but IMO it's a bit easier to follow.

Reply:
> By using __libcpp_datasizeof, we can get rid of __fits_in_tail_padding in this patch and IMO it makes the logic a bit easier to understand. What do you think?

Agreed, this would simplify the logic a bit! One downside would be that the `__union_t` would then always be potentially-overlapping (with a strict reading of the standard), even though, as you mentioned, it wouldn't make any difference to the layout. Then, when using `std::construct_at`/`std::destroy_at` on the `__union_t`, `std::launder` might have to be involved to make the code legal. Another thing is that GCC won't do guaranteed copy elision on potentially-overlapping subobjects, unlike clang. It will always fall back to a move constructor (this is bug <https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98995>).

So I would tend to keep the current "two-layout" version even if the code is more complex. Still, I have no strong feelings about this, as the two layouts are compatible as you said. Just GCC would fall back from guaranteed copy elision to move construction sometimes, I guess.

Reply:
Since this is extremely tricky code, I would tend to favour simplicity and ease of understanding above other concerns.

Reply:
In commit <https://github.com/llvm/llvm-project/pull/69673/commits/fe55535db87b8de529bd43a8f45c961f5c1dfbdd> I made the `__union_` member unconditionally `[[no_unique_address]]` to see if the CI complains. Sadly, GCC 13 fails the following tests:

```
  llvm-libc++-shared-gcc.cfg.in :: std/utilities/expected/expected.expected/ctor/ctor.convert.move.pass.cpp
  llvm-libc++-shared-gcc.cfg.in :: std/utilities/expected/expected.expected/ctor/ctor.copy.pass.cpp
  llvm-libc++-shared-gcc.cfg.in :: std/utilities/expected/expected.expected/ctor/ctor.move.pass.cpp
```

As GCC won't do guaranteed copy elision into a `[[no_unique_address]]` member it tries to fall back to move construction, but the move constructor of the `union_t` is deleted:

```
# | /home/runner/_work/llvm-project/llvm-project/build/generic-gcc/include/c++/v1/__expected/expected.h:115:9: error: use of deleted function ‘constexpr std::__1::__expected_base<_Tp, _Err>::__union_t::__union_t(std::__1::__expected_base<_Tp, _Err>::__union_t&&) [with _Tp = Data; _Err = int]’
# |   115 |       : __v(std::invoke(std::forward<_Func>(__f), std::forward<_Args>(__args)...)) {}
# |       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | /home/runner/_work/llvm-project/llvm-project/build/generic-gcc/include/c++/v1/__expected/expected.h:157:37: note: declared here
# |   157 |     _LIBCPP_HIDE_FROM_ABI constexpr __union_t(__union_t&&) = delete;
# |       |                                     ^~~~~~~~~
```

This is because in the current design the copy/move constructors of `expected` actually rely on guaranteed copy elision to work properly: <https://github.com/jiixyj/llvm-project/blob/fe55535db87b8de529bd43a8f45c961f5c1dfbdd/libcxx/include/__expected/expected.h#L323>

Of course, I could refactor this back to make use of `std::construct_at` instead of relying on guaranteed copy elision -- but then we would just trade one complexity for another... :/

Any suggestions or design ideas?